README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# USgrid <a href='https://ramikrispin.github.io/USgrid/'><img src='man/figures/USgrid.png' align="right"  /></a>

<!-- badges: start -->

[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html)
[![CRAN
status](https://www.r-pkg.org/badges/version/USgrid)](https://cran.r-project.org/package=USgrid)
[![License:
MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
<!-- badges: end -->

The USgrid R package provides a set of high frequency (hourly)
time-series datasets, describing the demand and generation of
electricity in the US (lower-48 states, excluding Alaska and Hawaii).
That includes the following series:

-   `US_elec` - the total hourly demand and supply (generation) for
    electricity in the US since July 2015

-   `US_source` - the US net generation of electricity by energy source
    (natural gas, coal, solar, etc.) since July 2018

-   `Cal_elec` - The California subregion hourly demand by operator
    since July 2018

All datasets are in [tsibble](https://tsibble.tidyverts.org/index.html)
format

**Source:** [US Energy Information
Administration](https://www.eia.gov/), Mar 2021

## Installation

Install the stable version from
[CRAN](https://CRAN.R-project.org/package=USgrid):

``` r
install.packages("USgrid")
```

or install the development version from
[Github](https://github.com/RamiKrispin/USgrid):

``` r
# install.packages("remotes")
remotes::install_github("RamiKrispin/USgrid")
```

## Examples

The hourly demand and generation (supply) of electricty in the US:

``` r
library(USgrid)
library(plotly)

data(US_elec)

plot_ly(data = US_elec,
        x = ~ date_time,
        y = ~ series,
        color = ~ type,
        colors = c("#66C2A5","#8DA0CB"),
        type = "scatter",
        mode = "lines") %>%
        layout(title = "US Electricity Demand vs. Supply (Hourly)",
               yaxis = list(title = "Mwh"),
               xaxis = list(title = "Source: US Energy Information Administration (Mar 2021)"))
```

<img src="man/figures/US_elec.svg" width="100%" />

The hourly generation (supply) of electricty in the US by source:

``` r
data("US_source")

plot_ly(data = US_source,
        x = ~ date_time,
        y = ~ series,
        color = ~ source,
        type = "scatter",
        mode = "lines") %>%
  layout(title = "US Electricity Generation by Energy Source",
         yaxis = list(title = "Mwh"),
         xaxis = list(title = "Source: US Energy Information Administration (Mar 2021)"))
```

<img src="man/figures/US_source.svg" width="100%" />

The California subregion hourly demand by operator

``` r
data("Cal_elec")

plot_ly(data = Cal_elec,
        x = ~ date_time,
        y = ~ series,
        color = ~ operator,
        type = "scatter",
        mode = "lines") %>%
  layout(title = "California Hourly Demand by Operator",
         yaxis = list(title = "Mwh"),
         xaxis = list(title = "Source: US Energy Information Administration (Mar 2021)"))
```

<img src="man/figures/Cal_elec.svg" width="100%" />


NEWS.md
# USgrid 0.1.2

* Updated the package datasets (up to date to March 15th, 2021)

# USgrid 0.1.1

* Updated the package datasets

# USgrid 0.1.0

Addws the following series:

* `US_elec`  - the total hourly demand and supply (generation) for electricity in the US since July 2015

* `US_source` - the hourly demand for electricity in the US by energy source (natural gas, coal, solar, etc.) since July 2018

* `Cal_elec` - The California subregion hourly demand by operator since July 2018

All series updated up to Dec 31 2019

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(USgrid)

test_check("USgrid")


tests/testthat/test-dataset.R
context("Test the datasets attributes")

test_that(desc = "Test US_elec dataset",
          {
            start_time <- base::min(US_elec$date_time)
            max_time <- base::max(US_elec$date_time)
            expect_equal(start_time == base::as.POSIXct("2015-07-01 05:00:00", tz = "UTC"), TRUE)
            expect_equal(max_time >= base::as.POSIXct("2019-11-24 23:00:00", tz = "UTC"), TRUE)
            expect_equal(tsibble::is_tsibble(US_elec), TRUE)
            expect_equal(tsibble::index(US_elec) == "date_time", TRUE)
            expect_equal(tsibble::key(US_elec) == "type", TRUE)
            expect_equal(lubridate::is.POSIXct(US_elec$date_time), TRUE)
          })


test_that(desc = "Test US_source dataset",
          {
            start_time <- base::min(US_source$date_time)
            max_time <- base::max(US_source$date_time)
            expect_equal(start_time == base::as.POSIXct("2018-07-01 05:00:00", tz = "UTC"), TRUE)
            expect_equal(max_time >= base::as.POSIXct("2019-11-24 23:00:00", tz = "UTC"), TRUE)
            expect_equal(tsibble::is_tsibble(US_source), TRUE)
            expect_equal(tsibble::index(US_source) == "date_time", TRUE)
            expect_equal(tsibble::key(US_source) == "source", TRUE)
            expect_equal(lubridate::is.POSIXct(US_source$date_time), TRUE)
          })


test_that(desc = "Test Cal_elec dataset",
          {
            start_time <- base::min(Cal_elec$date_time)
            max_time <- base::max(Cal_elec$date_time)
            expect_equal(start_time == base::as.POSIXct("2018-07-01 08:00:00", tz = "UTC"), TRUE)
            expect_equal(max_time >= base::as.POSIXct("2019-11-24 23:00:00", tz = "UTC"), TRUE)
            expect_equal(tsibble::is_tsibble(Cal_elec), TRUE)
            expect_equal(tsibble::index(Cal_elec) == "date_time", TRUE)
            expect_equal(tsibble::key(Cal_elec) == "operator", TRUE)
            expect_equal(lubridate::is.POSIXct(Cal_elec$date_time), TRUE)
          })


R/data.R
#' The US Hourly Demand and Supply for Electricity
#'
#' @description  The total hourly demand and supply (generation) for electricity in the US since July 2015
#'
#' Units: megawatthours
#'
#' Time zone: UTC
#'
#'
#' @format A tsibble object with hourly intervals
#' @source US Energy Information Administration (Mar 2021) \href{https://www.eia.gov/}{website}
#' @keywords datasets time-series
#' @details The dataset contains the hourly demand and supply (generation) for electricity in the US (megawatthours).
#' The `type` column describes the type of the series (demand or generation)
#' @examples
#'
#' data(US_elec)
#'
#' library(plotly)
#'
#' plot_ly(data = US_elec,
#'         x = ~ date_time,
#'         y = ~ series,
#'         color = ~ type,
#'         type = "scatter",
#'         mode = "lines")
#'

"US_elec"


#' The US Hourly Net Generation by Energy Source

#' @description The net generation of electricity in the US by energy source (i.e., natural gas, coal, solar, etc.) since July 2018.
#'
#' Units: megawatthours
#'
#' Time zone: UTC
#'
#'
#' @format A tsibble object with hourly intervals
#' @source US Energy Information Administration (Mar 2021) \href{https://www.eia.gov/}{website}
#' @keywords datasets time-series
#' @details The dataset contains the hourly net generation of electricity in the US (megawatthours) by energy source
#' The `source` column describes the type of the energy source
#' @examples
#'
#' data(US_source)
#'
#' library(plotly)
#'
#' plot_ly(data = US_source,
#'         x = ~ date_time,
#'         y = ~ series,
#'         color = ~ source,
#'         type = "scatter",
#'         mode = "lines")
#'

"US_source"

#' Demand for California Independent System Operator (CISO)

#' @description The demand for electricity in California subregion since July 2018 by operator.
#'
#' Units: megawatthours
#'
#' Time zone: UTC
#'
#'
#' @format A tsibble object with hourly intervals
#' @source US Energy Information Administration (Mar 2021) \href{https://www.eia.gov/}{website}
#' @keywords datasets time-series
#' @details The dataset contains the hourly demand for electricity in the California subregion (megawatthours).
#' The `operator` column describes the name of operator provider
#' @examples
#'
#' data(Cal_elec)
#'
#' library(plotly)
#'
#' plot_ly(data = Cal_elec,
#'         x = ~ date_time,
#'         y = ~ series,
#'         color = ~ operator,
#'         type = "scatter",
#'         mode = "lines")
#'

"Cal_elec"



data_raw/init_pull.R
#----------- Pulling the demand for electricity -----------

`%>%` <- magrittr::`%>%`

eia_series <- function(api_key, series_id){
  `%>%` <- magrittr::`%>%`

  url <- base::paste("http://api.eia.gov/series/?",
                     "api_key=", api_key,
                     "&series_id=", series_id,
                     "&out=json", sep = "")




  command <- base::paste("curl", " '",url, "' | jq -r '.series[].data[] | @tsv'", sep = "")

  output <- utils::read.table(text = system(command = command, intern = TRUE), sep = "\t") %>%
    stats::setNames(c("timestamp", "series")) %>%
    dplyr::mutate(date_time = lubridate::ymd_h(timestamp, tz = "UTC")) %>%
    dplyr::select(date_time, series) %>%
    dplyr::arrange(date_time)
  return(output)
}

#----------- Set a uniform end time for all series -----------
end_time <- as.POSIXct("2020-11-14 00:00:00", tz = "UTC") - lubridate::hours(1)
#----- Series 1 - Total Demand and Generation-----
source("/Users/ramikrispin/R/packages/APIs/eia.R")
tsAPI::eia_query(api_key = api_key,category_id = 2123635)


# US Demand
tsAPI::eia_query(api_key = api_key,category_id = 2122628)
tsAPI::eia_query(api_key = api_key,category_id = 3389935)

us_demand1<- eia_series(api_key = api_key, series_id  = "EBA.US48-ALL.D.H") %>%
  dplyr::mutate(type = "demand") %>%
  dplyr::arrange(date_time)

head(us_demand1)
tail(us_demand1)
table(is.na(us_demand1$series))

start_time <- NULL
start_time <- min(us_demand1$date_time)

us_demand <- data.frame(date_time = seq.POSIXt(from = start_time, to = end_time, by = "hour")) %>%
  dplyr::left_join(us_demand1,  by = "date_time")

# US Generation
tsAPI::eia_query(api_key = api_key,category_id = 2122629)
tsAPI::eia_query(api_key = api_key,category_id = 3390020)

us_gen1 <- eia_series(api_key = api_key, series_id  = "EBA.US48-ALL.NG.H") %>%
  dplyr::mutate(type = "generation") %>%
  dplyr::arrange(date_time)

head(us_gen1)

table(is.na(us_gen1$series))

start_time <- NULL
start_time <- min(us_gen1$date_time)

us_gen <- data.frame(date_time = seq.POSIXt(from = start_time, to = end_time, by = "hour")) %>%
  dplyr::left_join(us_gen1,  by = "date_time")

US_elec <- dplyr::bind_rows(us_demand, us_gen) %>%
  tsibble::as_tsibble(key = type, index = date_time)


head(US_elec)
tail(US_elec)

usethis::use_data(US_elec, overwrite = TRUE)

plotly::plot_ly(data = US_elec,
                x = ~ date_time,
                y = ~ series,
                color = ~ type,
                type = "scatter",
                mode = "lines")


#----- Demand by BA subregion - get catalog -----
tsAPI::eia_query(api_key = api_key,category_id = 2123635)
x <- tsAPI::eia_query(api_key = api_key,category_id = 3390016)

sub_df <- data.frame(category_ids = x$category$childcategories$category_id,
                     subregions = sub(" subregions", "",x$category$childcategories$name))

#----- Demand by BA subregion - function to pull the region data -----
reg_elec <- function(catalog, end_time){
  region_df <- lapply(1:nrow(catalog), function(i){
    df <- df1 <- df2 <- NULL
    print(catalog$series_id[i])
    df <- eia_series(api_key = api_key, series_id = catalog$series_id[i])

    if(!"error" %in% names(df$data)){

      df1 <- df %>%
        dplyr::arrange(date_time)

      start_time <- min(df1$date_time)

      df2 <- data.frame(date_time = seq.POSIXt(from = start_time, to = end_time, by = "hour")) %>%
        dplyr::left_join(df1, by = "date_time") %>%
        dplyr::mutate(operator = catalog$operator[i],
                      status = "ok")


    } else {

      #### Need to update ####
      warning(paste("The series", catalog$series_id[i], "return error"))
      df2 <- data.frame(date_time = NA,
                        series = NA,
                        operator = catalog$operator[i],
                        status = "error")
      #### Need to update ####
    }
    return(df2)
  }) %>% dplyr::bind_rows()

  region_temp <- region_df %>% dplyr::filter(status == "ok")

  if(nrow(region_temp) == 0){
    stop("No valid observations were found")
  } else {
    region_total <- region_temp %>%
      dplyr::group_by(date_time) %>%
      dplyr::summarise(series = sum(series, na.rm = any(!is.na(series)))) %>%
      dplyr::mutate(operator = "Total",
                    status = "ok")

    region_df <- dplyr::bind_rows(region_df, region_total)
  }
  return(region_df)
}



#----- Demand by BA subregion - California -----
cal_cat <- tsAPI::eia_query(api_key = api_key,category_id = 3390291)$category$childseries %>%
  dplyr::mutate(flag = grepl("UTC time", name)) %>%
  dplyr::filter(flag) %>%
  dplyr::select(-flag) %>%
  dplyr::mutate(operator = trimws(sapply(strsplit(name, split = ","),function(x) x[length(x) - 1])))

Cal_elec <- reg_elec(catalog = cal_cat, end_time = end_time)
table(Cal_elec$status)

if(all(Cal_elec$status == "ok")){
  Cal_elec <- Cal_elec %>%
    dplyr::select(-status) %>%
    tsibble::as_tsibble(key = operator, index = date_time)
} else {
  warning("Some observations do not have 'ok' status, check the series")
}

plotly::plot_ly(data = Cal_elec,
                x = ~ date_time,
                y = ~ series,
                color = ~ operator,
                type = "scatter",
                mode = "lines")


usethis::use_data(Cal_elec, overwrite = TRUE)

#----- Net generation by energy source -----

energy_source_cat <- tsAPI::eia_query(api_key = api_key,category_id = 3390105)$category$childseries %>%
  dplyr::mutate(flag = grepl("UTC time", name)) %>%
  dplyr::filter(flag) %>%
  dplyr::select(-flag) %>%
  dplyr::mutate(operator = lapply(name, function(i){
    trimws(strsplit(strsplit(i, split = c("from"))[[1]][2], split = "for")[[1]][1])}) %>%
      unlist)


US_source <- reg_elec(catalog = energy_source_cat, end_time = end_time)



if(all(US_source$status == "ok")){
  US_source <- US_source %>%
    dplyr::filter(operator != "Total") %>%
    dplyr::select(date_time, series, source = operator) %>%
    tsibble::as_tsibble(key = source, index = date_time)
} else {
  warning("Some observations do not have 'ok' status, check the series")
}

head(US_source)

plotly::plot_ly(data = US_source,
                x = ~ date_time,
                y = ~ series,
                color = ~ source,
                type = "scatter",
                mode = "lines")

usethis::use_data(US_source, overwrite = TRUE)


data_raw/data_refresh.R
#----------- Data refresh function -----------

dt_refresh <- function(end_time, api_key = Sys.getenv("eia_key")){

  end_time_old <- us_elec_old <- us_elec_old1 <- NULL
  us_source_old <- us_source_old1 <- us_source_new <- NULL
  cal_elec_old <- cal_elec_old1 <- cal_elec_new <-NULL
  start_time <- NULL
  # Error handling
  if(!lubridate::is.POSIXct(end_time)){
    stop("The 'end_time' argument is not POSIXct")
  }

  #--- Set functions ---#

  # Add pipe function
  `%>%` <- magrittr::`%>%`

  # Set function to pull data from eia API
  eia_series <- function(api_key, series_id){

    url <- base::paste("http://api.eia.gov/series/?",
                       "api_key=", api_key,
                       "&series_id=", series_id,
                       "&out=json", sep = "")
    # Parse the data with jq
    command <- base::paste("curl", " '",url, "' | jq -r '.series[].data[] | @tsv'", sep = "")

    output <- utils::read.table(text = system(command = command, intern = TRUE), sep = "\t") %>%
      stats::setNames(c("timestamp", "series")) %>%
      dplyr::mutate(date_time = lubridate::ymd_h(timestamp, tz = "UTC")) %>%
      dplyr::select(date_time, series) %>%
      dplyr::arrange(date_time)
    return(output)
  }

  # Pulling multiple series by series catalog number
  reg_elec <- function(catalog, end_time){
    region_df <- lapply(1:nrow(catalog), function(i){
      df <- df1 <- df2 <- NULL
      print(catalog$series_id[i])
      df <- eia_series(api_key = api_key, series_id = catalog$series_id[i])

      if(!"error" %in% names(df$data)){

        df1 <- df %>%
          dplyr::arrange(date_time)

        start_time <- min(df1$date_time)

        df2 <- data.frame(date_time = seq.POSIXt(from = start_time, to = end_time, by = "hour")) %>%
          dplyr::left_join(df1, by = "date_time") %>%
          dplyr::mutate(operator = catalog$operator[i],
                        status = "ok")


      } else {

        #### Need to update ####
        warning(paste("The series", catalog$series_id[i], "return error"))
        df2 <- data.frame(date_time = NA,
                          series = NA,
                          operator = catalog$operator[i],
                          status = "error")
        #### Need to update ####
      }
      return(df2)
    }) %>% dplyr::bind_rows()

    region_temp <- region_df %>% dplyr::filter(status == "ok")

    if(nrow(region_temp) == 0){
      stop("No valid observations were found")
    } else {
      region_total <- region_temp %>%
        dplyr::group_by(date_time) %>%
        dplyr::summarise(series = sum(series, na.rm = any(!is.na(series)))) %>%
        dplyr::mutate(operator = "Total",
                      status = "ok")

      region_df <- dplyr::bind_rows(region_df, region_total)
    }
    return(region_df)
  }


  #--- load saved datasets ---#
  us_elec_old <- USgrid::US_elec
  us_source_old <- USgrid::US_source
  cal_elec_old <- USgrid::Cal_elec

  end_time_old <- base::min(base::max(us_elec_old$date_time),
                            base::max(us_source_old$date_time),
                            base::max(cal_elec_old$date_time))

  start_time <- end_time_old + lubridate::hours(1)


  if(start_time > end_time){
    stop("The 'start_time' is greater than the 'end_time")
  }


  #--- Pulling US_elec series ---#
  us_demand1<- eia_series(api_key = api_key, series_id  = "EBA.US48-ALL.D.H") %>%
    dplyr::mutate(type = "demand") %>%
    dplyr::arrange(date_time)
  us_elec_start <- base::min(us_elec_old$date_time)
  us_demand <- data.frame(date_time = seq.POSIXt(from = us_elec_start, to = end_time, by = "hour")) %>%
    dplyr::left_join(us_demand1,  by = "date_time")


  us_gen1 <- eia_series(api_key = api_key, series_id  = "EBA.US48-ALL.NG.H") %>%
    dplyr::mutate(type = "generation") %>%
    dplyr::arrange(date_time)

  head(us_gen1)


  us_gen <- data.frame(date_time = seq.POSIXt(from = us_elec_start, to = end_time, by = "hour")) %>%
    dplyr::left_join(us_gen1,  by = "date_time")

  us_elec_new <- dplyr::bind_rows(us_demand, us_gen) %>%
    tsibble::as_tsibble(key = type, index = date_time)


  # Valdidate the dataset

  us_elec_old1 <- us_elec_new %>% dplyr::filter(date_time <= end_time_old)

  if(!base::identical(us_elec_old, us_elec_old1)){
    stop("It seems like the new pull is not match to the old one for the 'US_elec' series")
  }

  #--- Pulling US_source series ---#
  energy_source_cat <- tsAPI::eia_query(api_key = api_key,category_id = 3390105)$category$childseries %>%
    dplyr::mutate(flag = grepl("UTC time", name)) %>%
    dplyr::filter(flag) %>%
    dplyr::select(-flag) %>%
    dplyr::mutate(operator = lapply(name, function(i){
      trimws(strsplit(strsplit(i, split = c("from"))[[1]][2], split = "for")[[1]][1])}) %>%
        unlist)


  us_source_new <- reg_elec(catalog = energy_source_cat, end_time = end_time)



  if(base::all(us_source_new$status == "ok")){
    us_source_new <- us_source_new %>%
      dplyr::filter(operator != "Total") %>%
      dplyr::select(date_time, series, source = operator) %>%
      tsibble::as_tsibble(key = source, index = date_time)
  } else {
    warning("Some observations do not have 'ok' status, check the series")
  }

  # Valdidate the dataset
  us_source_old1 <- us_source_new %>% dplyr::filter(date_time <= end_time_old)

  if(!base::identical(us_source_old, us_source_old1)){
    stop("It seems like the new pull is not match to the old one for the 'US_source' series")
  }

  #--- Pulling Cal_elec series ---#

  cal_cat <- tsAPI::eia_query(api_key = api_key,category_id = 3390291)$category$childseries %>%
    dplyr::mutate(flag = grepl("UTC time", name)) %>%
    dplyr::filter(flag) %>%
    dplyr::select(-flag) %>%
    dplyr::mutate(operator = trimws(sapply(strsplit(name, split = ","),function(x) x[length(x) - 1])))

  cal_elec_new <- reg_elec(catalog = cal_cat, end_time = end_time)

  if(all(cal_elec_new$status == "ok")){
    cal_elec_new <- cal_elec_new %>%
      dplyr::select(-status) %>%
      tsibble::as_tsibble(key = operator, index = date_time)
  } else {
    warning("Some observations do not have 'ok' status, check the series")
  }

  # Valdidate the dataset
  cal_elec_old1 <- cal_elec_new %>% dplyr::filter(date_time <= end_time_old)

  if(!base::identical(cal_elec_old, cal_elec_old1)){
    stop("It seems like the new pull is not match to the old one for the 'Cal_elec' series")
  }


  #--- Saving the series ---#
  # US_elec
  base::print(plotly::plot_ly(data = us_elec_new,
                  x = ~ date_time,
                  y = ~ series,
                  color = ~ type,
                  type = "scatter",
                  mode = "lines"))


  base::cat(base::paste(" US_elec series summary", "\n",
                        "----------------------", "\n",
                        "Start time:", base::min(us_elec_new$date_time), "\n",
                        "End time:", base::max(us_elec_new$date_time), "\n",
                        "Missing values:", base::any(base::is.na(us_elec_new$series))))


  q <- readline("Do you want to save the updated US_elec series? [Y/n]")
  if(q == "" | base::tolower(q) == "y"){
    US_elec <- us_elec_new
    usethis::use_data(US_elec, overwrite = TRUE)
  }

  # US_source
  base::print(plotly::plot_ly(data = us_source_new,
                  x = ~ date_time,
                  y = ~ series,
                  color = ~ source,
                  type = "scatter",
                  mode = "lines"))


  base::cat(base::paste(" US_source series summary", "\n",
                        "-------------------------", "\n",
                        "Start time:", base::min(us_source_new$date_time), "\n",
                        "End time:", base::max(us_source_new$date_time), "\n",
                        "Missing values:", base::any(base::is.na(us_source_new$series))))


  q <- readline("Do you want to save the updated US_source series? [Y/n]")
  if(q == "" | base::tolower(q) == "y"){
    US_source <- us_source_new
    usethis::use_data(US_source, overwrite = TRUE)
  }

  # Cal_elec
  base::print(plotly::plot_ly(data = cal_elec_new,
                  x = ~ date_time,
                  y = ~ series,
                  color = ~ operator,
                  type = "scatter",
                  mode = "lines"))

  base::cat(base::paste(" Cal_elec series summary", "\n",
                        "-----------------------", "\n",
                        "Start time:", base::min(cal_elec_new$date_time), "\n",
                        "End time:", base::max(cal_elec_new$date_time), "\n",
                        "Missing values:", base::any(base::is.na(cal_elec_new$series))))


  q <- readline("Do you want to save the updated Cal_elec series? [Y/n]")
  if(q == "" | base::tolower(q) == "y"){
    Cal_elec <- cal_elec_new
    usethis::use_data(Cal_elec, overwrite = TRUE)
  }



}



dt_refresh(end_time = lubridate::ymd_hms("2021-03-15 00:00:00", tz = "UTC"))



